Add SEPA Direct Debit payment method.

superdav42 opened the pull request:
Fixes #1468 

Adds the SEPA direct debit payment method. Is hidden behind a feature flag controlled by the `_wcpay_feature_sepa` option.
This MVP implementation and there will be many follow up PRs. This is being merged now to keep the PR smaller.

### Testing instructions
* set store currency to euro
* Enable feature flag wp option update _wcpay_feature_sepa 1
* Enable sepa in checkbox at bottom of /wp-admin/admin.php?page=wc-settings&tab=checkout&section=woocommerce_payments
* Add item to cart and go to checkout
* choose SEPA direct debit payment option
* enter test IBAN DE89370400440532013000
* Place order
* Order should be placed and show the PAYMENT METHOD: SEPA Direct Debit
* Order status should be set to on-hold

### Know Issues to be address seperately
[Transaction details page for SEPA does not load in wcpay transaction tab
](https://github.com/Automattic/woocommerce-payments/issues/1472 )
[Saved CC will be used if selected instead of SEPA even if it's selected
](https://github.com/Automattic/woocommerce-payments/issues/1471)
[Unable to pay for an existing order using SEPA because it says customer email is not provided.
](https://github.com/Automattic/woocommerce-payments/issues/1539)
[SEPA method does not appear on block checkout
](https://github.com/Automattic/woocommerce-payments/issues/1470)
JS code could be better architectures when dealing with 3+ payment methods.

kalessil commented on includes/payment-method/class-card.php:
```diff
@@ -0,0 +1,18 @@
+ * @package WCPay\Payment_Gateway
+ */
+
+namespace WCPay\Payment_Method;
```
Using `Gateways` instead of `Payment_Method` would better express intention here.

superdav42 replied:
There are already other opinions on the name in this conversation p1617864347114400/1617748545.084400-slack-CGGCLBN58 . Maybe we should put it to a vote? I really don't have an opinion.

kalessil replied:
Sounds good, I'd suggest using wcpay channel for the voting. We can pack the changes based on voting into a follow-up PR as well.

kalessil commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -29,6 +29,8 @@ class WC_Payment_Gateway_WCPay extends WC_Payment_Gateway_CC {
 	 */
 	const GATEWAY_ID = 'woocommerce_payments';
 
+	const METHOD_ENABLED_KEY = 'enabled';
```
`GATEWAY_ENABLED_KEY` would be more expressive.

kalessil commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -745,7 +758,7 @@ public function process_payment_for_order( $cart, $payment_information ) {
 
 		if ( ! empty( $intent ) ) {
-			if ( 'succeeded' !== $status && 'requires_capture' !== $status ) {
+			if ( 'succeeded' !== $status && 'requires_capture' !== $status && 'processing' !== $status ) {
```
Let's use `in_array()` here.

kalessil commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -1632,7 +1682,7 @@ public function update_order_status() {
 			$order->save();
 
-			if ( 'succeeded' === $status || 'requires_capture' === $status ) {
+			if ( 'succeeded' === $status || 'requires_capture' === $status || 'processing' === $status ) {
```
Let's use in_array() here.

kalessil commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -1075,13 +1089,48 @@ public function generate_account_statement_descriptor_html( $key, $data ) {
 		}
 	}
 
+	/**
```
I had the impression that the new method introduced in order to make `\WC_Payment_Gateway_WCPay::is_enabled`, `\WC_Payment_Gateway_WCPay::disable`, `\WC_Payment_Gateway_WCPay::enable` for the new gateways.

I suggest simplifying this:
- introduce the setting key as a constant (no dynamic calculation)
- refactor the methods to use the constant
- override the constant for each gateway

That would make this PR even more awesome =)

superdav42 replied:
I think I'll let the setting squad handle further iterations around this. For now I want the main WC_Payment_Gateway_WCPay gateway to be responsible for managing all setting so they will stay in the same option key and not need migrating. There will be payment gateway specific setting but for now all should use the same. Eventually I think we want an option key for common setting and an option key for each gateway's settings.

kalessil replied:
Sounds good.

kalessil reviewed (changes requested):
Hey @superdav42, thanks for bringing SEPA support into WCPay: great job!

I tested the PR and it works as described (also I see a list of known issues, thanks for mentioning them). 

I left several comments we need to address on this review iteration. Didn't look into test cases yest, as have to dive into the PR, but will get to them once the PR is ready for the next review.

superdav42 commented on includes/payment-method/class-card.php:
```diff
@@ -0,0 +1,18 @@
+ * Credit Card Payment method.
+ * Right now behaves exactly like WC_Payment_Gateway_WCPay for max compatibility.
+ */
+class Card extends WC_Payment_Gateway_WCPay {
```
Making the base class abstract would require more code changes in the unit tests and I was trying to keep this PR small but it's something that should probably be done eventually. I think when we add the 3rd payment method it will become more clear the structure that would work best.

kalessil replied:
Agree, let's keep those as it is.

superdav42 commented:
@kalessil Thanks for your review. I addressed most of your concerns. I think we may want to bring more team members into the decision on which namespace and naming convention to use for new gateways. I think the other suggestions would be best to wait for a future PR.

kalessil commented:
Thank you for the fast update @superdav42 - the changes are looking great, giving the PR as a final review.

PS: we can surely shift comments related to namespaces, options, and gateway availability checks into a follow-up PR.

kalessil reviewed (approved):
LGTM, :shipit: 

And the list of topics for a follow-up PR, so we don't loose the track of them:
- gateways namespaces voting
- sepa gateway availability for EUR order only
- options handling

superdav42 commented:
Created some follow up issues:
EUR currency check will be added in https://github.com/Automattic/woocommerce-payments/issues/1526
Options improvements in https://github.com/Automattic/woocommerce-payments/issues/1576
Naming things vote in https://github.com/Automattic/woocommerce-payments/issues/1581
